### PR TITLE
Modified UserFateStore.create to behave like MetaFateStore

### LINKED ISSUE #4246: Move the test case in AccumuloStoreIT into FateStoreIT
In #4160 an IT was added to test some of the conditions that were added to the conditional mutations. As mentioned below, it would be nice to move this test case to also run against zookeeper.

> This is a nice test.  If possible would be nice to eventually run the test against Zookeeper also like #4202.  If it makes sense could be a follow on.

_Originally posted by @keith-turner in https://github.com/apache/accumulo/pull/4160#discussion_r1470246756_

### PR DESCRIPTION
**Pull request by dlmarion**
MetaFateStore.create will retry forever when a collision happens when trying to create a Fate transaction. The probabiliy of a collision is low due to a random UUID being used. Before this change the UserFateStore.create method would retry 5 times then throw an exception. This was removed in favor of an unlimited retry so that the behavior of the two Fate stores is the same.

Closes #4246

### CODE REVIEW

**Review by kevinrr888 — APPROVED** [review]
This is a good change and keeps the two stores consistent. Wasn't sure how this would conflict with the changes in https://github.com/apache/accumulo/pull/4524, but looking at this, shouldn't have any problems.

**Comment by kevinrr888** [comment]
Hmm thinking about this some more. Wondering if it would be better if we had finite number of retries for both stores instead of infinite. The probability of collision is very low. 

I'm not sure that both stores are functioning the same under these new changes. The MFS will throw an error if any other KeeperException is seen other than NodeExistsException. The UFS will retry forever if an UNKNOWN status is received. So if there is something wrong and we keep receiving UNKNOWN, it will keep retrying.

**Comment by dlmarion** [comment]
> Hmm thinking about this some more. Wondering if it would be better if we had finite number of retries for both stores instead of infinite. The probability of collision is very low.
> 
> I'm not sure that both stores are functioning the same under these new changes. The MFS will throw an error if any other KeeperException is seen other than NodeExistsException. The UFS will retry forever if an UNKNOWN status is received. So if there is something wrong and we keep receiving UNKNOWN, it will keep retrying.

So, I first went down the path of creating an IT to determine what error a user would receive when something like `TableOperations.create` failed when the number of retries in the UserFateStore would be exceeded. When I realized that MetaFateStore tried infinitely, I looked at how we could test for a duplicate FateId in the ConditionalMutationWriter. I didn't see an easy way to do that since the FateId is in the `row`.  If we could do that, then we could set a status for duplicate FateId, and continue to retry, and then fail on REJECTED or UNKNOWN.

**Comment by kevinrr888** [comment]
I'm just wondering if it would be better/safer to keep a finite number of retries and maybe change MFS to be a finite number of retries as well. It should still retry on UNKNOWN or REJECTED for UFS

**Comment by kevinrr888** [comment]
My comment regarding the retry for UNKNOWN was just that if there is something wrong with the TabletServer or something else that would cause UNKNOWN to be received indefinitely, this current impl will retry forever.

**Comment by dlmarion** [comment]
I think with the changes in 5882212, the two implementations are consistent. They continue forever when there is a dupe, otherwise they throw an exception.

We can have a discussion about whether or not they should try forever and I'm wondering what they did previously. Because I think that's the answer as the calling code is already written to handle that.

File: core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java [hunk 1]
```diff
@@ -105,14 +106,17 @@ public FateId create() {
         case ACCEPTED:
           return fateId;
         case UNKNOWN:
+          log.error("Fate creation returned UNKNOWN status, raising exception");
+          throw new IllegalStateException("Fate creation returned UNKNOWN status");
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:110]: I don't think we want to throw an exception in this case. This just means it's unknown if the mutation was successfully written or not, in which case we would want to just retry. The UNKNOWN status isn't quite equivalent to the KeeperException thrown in MFS

**Comment by kevinrr888** [comment]
Apologies if any of my comments were confusing/misleading.
My only concern is retrying infinitely, not about the logic for retrying. It may be perfectly okay to retry infinitely, I'm just not sure and was throwing that out for discussion

File: core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java (hunk 1 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:110]: Ok, I looked further at ConditionalWriter.Status.UNKNOWN and how/where it's used. I agree with your assessment and can back this out.

**Comment by dlmarion** [comment]
> Apologies if any of my comments were confusing/misleading. My only concern is retrying infinitely, not about the logic for retrying. It may be perfectly okay to retry infinitely, I'm just not sure and was throwing that out for discussion

All good, I'm just trying to make sure things are handled consistently. Based on the old ZooStore implementation, which is now the MetaFateStore class, I think we should retry indefinitely in the UserFateStore. Reason being that the code that calls `Fate.startTransaction()` doesn't currently expect or handle an exception that would cause them to retry. The code that calls `Fate.startTransaction` expects success. If we want to limit the number of retries, then we should throw a `RetriesExceededException` or similar and modify the calling code to either retry or supply appropriate messaging to the user that their call to `TableOperations.compact` or similar has failed to start.

**Review by cshannon — APPROVED** [review]
This seems like a good change to me so things are consistent. I think it makes sense to treat both stores the same and keep retrying forever for the normal `create()` case where we are generating a UUID (vs supplying it).

As already noted, the meta store is pretty much guaranteed to eventually succeed because it just generates new UUIDs on a collision which is very rare anyways. It will still abort and exit if an exception is thrown due to an issue with Zookeeper (such as the client can't get to the server or something).

For the User store, it also generates a random ID in the same way so it should be fine to keep retrying and I don't really see an issue with that. I think we can keep retrying for `REJECTED` for sure and probably `UNKNOWN`.  

1. `REJECTED` of course just means there was a collision so we definitely want to retry with a new UUID. 
2. For the `UNKNOWN` case that might mean there was just a transient issue with the client getting confirmation after submission so we don't know if it went through so I think we can try again without issue. If it was actually successful the FateCleaner will handle cleanup and age off. 
3. If another type of AccumuloException is thrown (like the client can't connect at all anymore) or security exception is thrown then a runtime exception [bubbles](https://github.com/dlmarion/accumulo/blob/99b1eced6cd4f6c1da35bf039fbff4c40e5e0171/core/src/main/java/org/apache/accumulo/core/fate/user/FateMutatorImpl.java#L211-L217) up and we will also exit the method (just like with the meta store and a ZK error).

The time when we should limit the retries would be when a FateKey is provided and we are generating the ID off of the FateKey. That [still](https://github.com/dlmarion/accumulo/blob/99b1eced6cd4f6c1da35bf039fbff4c40e5e0171/core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java#L126) has a retry of 5 and that makes sense so you don't block forever with multiple attempts as that ID will keep trying the same one over and over and won't generate a new one.

@keith-turner - Do see you any issue with this approach?

**Review by keith-turner — APPROVED** [review]

**Comment by keith-turner** [comment]
> @keith-turner - Do see you any issue with this approach?

No, that all sounds good.

**Comment by keith-turner** [comment]
> For the UNKNOWN case that might mean there was just a transient issue with the client getting confirmation after submission so we don't know if it went through so I think we can try again without issue. If it was actually successful the FateCleaner will handle cleanup and age off.

More complexity could be added around the unknown case, but its not worth the maintenance burden. I agree its better to just let the FateCleaner deal with it.

**Comment by cshannon** [comment]
> More complexity could be added around the unknown case, but its not worth the maintenance burden. I agree its better to just let the FateCleaner deal with it.

Yeah I figured it wouldn't be worth trying to scan and check if it was successful as it should be rare and the FateCleaner would clean it up. To really handle UNKNOWN effectively you'd need something more complex like how Ample combines a retry and rejection handler API.

**Review by kevinrr888 — APPROVED** [review]

**Pull request merged**